lib/posix/options/pipe.c

```diff
@@ -0,0 +1,343 @@
+{
+	size_t avail = ring_buf_space_get(&p->buf);
+
+	size = MIN(size, avail);
+	ring_buf_put(&p->buf, data, size);
+	avail = ring_buf_space_get(&p->buf);
+
+	if (avail == 0) {
+		k_poll_signal_reset(&p->write_sig);
+	}
+
+	k_poll_signal_raise(&p->read_sig, 0);
+
+	return size;
+}
+
+static ssize_t pipe_read_op(void *obj, void *buf, size_t size)
+{
+	pipe_rw_op_t op = pipe_read_locked;
+
+	return pipe_rw_op(obj, buf, size, op);
+}
+
+static ssize_t pipe_write_op(void *obj, const void *buf, size_t size)
+{
+	pipe_rw_op_t op = pipe_write_locked;
+
+	return pipe_rw_op(obj, (void *)buf, size, op);
+}
+
+static int pipe_close_op(void *obj)
+{
+	struct k_mutex *lock = NULL;
+	struct k_condvar *cond = NULL;
+	struct pipe_obj *p = (struct pipe_obj *)obj;
+
+	if (k_is_in_isr()) {
+		errno = EWOULDBLOCK;
+		return -1;
+	}
+
+	if (!z_get_obj_lock_and_cond(obj, &pipe_read_vtable, &lock, &cond)) {
+		if (!z_get_obj_lock_and_cond(obj, &pipe_write_vtable, &lock, &cond)) {
+			__ASSERT(false, "z_get_obj_lock_and_cond() failed");
+		}
+	}
+
+	k_mutex_lock(lock, K_FOREVER);
+	k_mem_slab_free(&pipe_slab, &p->buf.buffer);
+	/* when closing a pipe, broadcast to all waiters */
+	k_condvar_broadcast(cond);
+	k_mutex_unlock(lock);
+
+	return 0;
+}
+
+static int pipe_ioctl_op(void *obj, unsigned int request, va_list args)
+{
+	int ret;
+	struct pipe_obj *p = (struct pipe_obj *)obj;
+
+	K_SPINLOCK(&p->lock) {
+		switch (request) {
+		case F_GETFL:
+			ret = p->flags;
+			break;
+
+		case F_SETFL: {
+			int flags;
+
+			flags = va_arg(args, int);
+			p->flags |= flags;
+			ret = 0;
+		} break;
+
+		case ZFD_IOCTL_POLL_PREPARE: {
+			struct zsock_pollfd *pfd;
+			struct k_poll_event **pev;
+			struct k_poll_event *pev_end;
+
+			pfd = va_arg(args, struct zsock_pollfd *);
+			pev = va_arg(args, struct k_poll_event **);
+			pev_end = va_arg(args, struct k_poll_event *);
+
+			ret = pipe_poll_prepare(obj, pfd, pev, pev_end);
+		} break;
+
+		case ZFD_IOCTL_POLL_UPDATE: {
+			struct zsock_pollfd *pfd;
+			struct k_poll_event **pev;
+
+			pfd = va_arg(args, struct zsock_pollfd *);
+			pev = va_arg(args, struct k_poll_event **);
+
+			ret = pipe_poll_update(obj, pfd, pev);
+		} break;
+
+		default:
+			errno = EOPNOTSUPP;
+			ret = -1;
+			break;
+		}
+	}
+
+	return ret;
+}
+
+static const struct fd_op_vtable pipe_read_vtable = {
+	.read = pipe_read_op,
+	.close = pipe_close_op,
+	.ioctl = pipe_ioctl_op,
+};
+
+static const struct fd_op_vtable pipe_write_vtable = {
+	.write = pipe_write_op,
+	.close = pipe_close_op,
+	.ioctl = pipe_ioctl_op,
+};
+
+/* common to both pipe_read_op() and pipe_write_op() */
+static ssize_t pipe_rw_op(void *obj, void *buf, size_t size, pipe_rw_op_t op)
+{
+	int err;
+	ssize_t ret;
+	k_spinlock_key_t key;
+	struct pipe_obj *p = obj;
+	struct k_mutex *lock = NULL;
+	struct k_condvar *cond = NULL;
+
+	if (buf == NULL) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	key = k_spin_lock(&p->lock);
```

Comment:
Given that the two sides may have different blocking flags, I think you want to unify the locking here, presumably using a mutex, so that you don't have any mysterious races. Will end up simpler and easier to review.